apollo-federation/src/sources/connect/json_selection/helpers.rs

```diff
@@ -78,6 +78,16 @@ pub fn json_type_name(v: &JSON) -> &str {
     }
 }
 
+pub fn immutable_vec_push<T>(vec: &[T], item: T) -> Vec<T>
+where
+    T: Clone,
+{
+    let mut new_vec = Vec::with_capacity(vec.len() + 1);
+    new_vec.extend_from_slice(vec);
+    new_vec.push(item);
+    new_vec
```

Comment:
I'd prefer we just extend the vec directly, if we can. I _believe_ we never want duplicates of these, so it should be a problem to hand around ownership.

Passing `mut Vec<blah>` doesn't carry the same risks as `&mut Vec<blah>`, since you're forced to still return the new value for consumers to get the updates—so the functions are still "pure"

Reply:
It doesn't seem like there's already a `Vec` method to push an item and return the same (owned) vector, so this helper function still exists but has now become very short:
```rust
pub fn vec_push<T>(mut vec: Vec<T>, item: T) -> Vec<T> {
    vec.push(item);
    vec
}
```